app/multitenant/dynamo_collector.go

```diff
@@ -485,10 +485,21 @@ func (c *dynamoDBCollector) WaitOn(ctx context.Context, waiter chan struct{}) {
 	c.waitersLock.Unlock()
 
 	go func() {
-		_, err := sub.NextMsg(natsTimeout)
-		natsRequests.WithLabelValues("NextMsg", errorCode(err)).Add(1)
-		log.Debugf("NextMsg error: %v", err)
-		close(waiter)
+		for {
+			_, err := sub.NextMsg(natsTimeout)
+			if err == nats.ErrTimeout {
+				continue
+			}
+			natsRequests.WithLabelValues("NextMsg", errorCode(err)).Add(1)
+			if err != nil {
+				log.Debugf("NextMsg error: %v", err)
+				return
+			}
+			select {
+			case waiter <- struct{}{}:
+			default:
+			}
+		}
```

Comment:
Optional: Since you made this change because you discovered a bug, it'd be nice if this could have some sort of test that fails with the previous version & passes with this one.  

I can see why that might be a bit of a faff though.